### DISCUSSION TITLE
Best practice where should I put a WebSocket Connection with Recoil?

### DISCUSSION THREAD

**Discussion by SmashPlusUltra** [facebookexperimental/Recoil#1701] [Q&A] [answered]
If I'm using react as it is I expose the sockect connection using useContext, if I'm using Redux I do it in a Middleware, but now that I'm starting with recoil I don't know where to put it. Using useContext would be ok or if there's a suggestion would it should be using recoil?.

Any thoughts would be appreciated. 

---------
My try, I'm using signalr, the object to save looks like the following:
![hubConnectionObject](https://user-images.githubusercontent.com/53838441/160546213-d34271d5-f63d-4183-8925-0a539b02d05a.jpg)

```
export const hubConnection = atom({ key: 'hubConnection', default: false })
const worker = selector({
    key: 'worker',   
    get: ({ get }) => {
        return get(hubConnection )
    },
    set: ({ get, set }, op)=>{
        switch (op) {
            default:
                break;
        }
    }
  
})
const connect = selector({
    key: 'conn',
    get: ({ get }) => {
        return get(isLogIn)
    },
    set: ({ get, set }, op) => {
        const webSocketRouter = 'wss://localhost:5002/chathub'
        const currentUserid ='admin'
       set(hubConnection , new HubConnectionBuilder()
                    .withUrl(webSocketRouter + `?username=${currentUserid }`)
                    .withAutomaticReconnect()
                    .build())   
}
})
```
I got...
Uncaught (in promise) TypeError: Cannot redefine property: info
    at Function.defineProperties (<anonymous>)
    at disableLogs (react-dom.development.js:789:1)
    at getStateFromUpdate (react-dom.development.js:12106:1)
    at processUpdateQueue (react-dom.development.js:12250:1)
    at updateClassInstance (react-dom.development.js:13013:1)....

**Comment by drarmstr** [comment, marked as answer]
Perhaps [atom effects](https://recoiljs.org/docs/guides/atom-effects) would be appropriate?

**Comment by Jynx** [comment]
@SmashPlusUltra Were you able to create a full working solution here using an atom effect? I am interested in using signalR in tandem with recoil and am tinkering through the initial hubConnection state right now.

**Reply by SmashPlusUltra** [reply]
I didn't implement it in that way, I've moved my signalR setup to a shared worker in order to see if I could gain more freeing the main thread from processing the incoming messages. So I post messages to other objects witch in turn update atoms - run selectors etc.

**Reply by Jynx** [reply]
I see. And when you say a worker, you're talking about a service worker, and perhaps something like this? https://github.com/dotnet/aspnetcore/blob/9158039c9515e46d39c90f462ae91827769889c4/src/SignalR/samples/SignalRSamples/wwwroot/worker.js

Appreciate the response(s)!

**Reply by SmashPlusUltra** [reply]
Yes, exactly... particularly I've used a shared worker https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker

**Reply by Jynx** [reply]
Excellent. I'll look into this as well.

Out of curiosity, were you able to get a working recoil only solution to configuring and connecting to the hub before deciding to move to a shared worker solution?

**Reply by SmashPlusUltra** [reply]
Never had it working, I've tried but my problem change and my needs as well... so I ended up with the solution I comment before. I may say my project uses Recoil-Redux-UseContext 😂 almost everything under the sun.
